### ISSUE TITLE
Configure interactions options depending on the phenotype

### ISSUE THREAD

**Issue by vmt25** [pombase/canto#1991]
(Making sub-ticket of #1962, as agreed on the 24/07/2019 meeting)

Depending on the type of phenotypes some types of interactions are a bit nonsensical.
Consider the example of 'viable', which is a 'normal phenotype':
a) 'mut X viable, enhanced by mut B' - makes no sense because the mutant is already viable
mut X viable, suppressed by mut B - this probably means becoming partially lethal or lethal, which is a synthetic phenotype and should be annotated as a multi-gene phenotype.

Therefore, it would be good to an interaction config file (similar to the extension_config.tsv) for terms and sub-tree with restrictions:
CV term (and child terms), annotation, allowed interaction type

**Comment by kimrutherford** [comment]
I think we should discuss this next time on Skype.  I can think of several ways I could get this wrong so I'd like to get it clear in my head before starting.  :-)

**Comment by kimrutherford** [comment]
Val, Midori and Antonia, will this work for PomBase too?

**Comment by mah11** [comment]
> will this work for PomBase too?

At first glance, it seems like it would probably be useful sooner or later. It does make sense to configure Canto not to show interaction types that wouldn't make sense for the phenotype.

I have a question - or maybe a couple of related questions - though. In the list of config fields:

> CV term (and child terms), annotation, allowed interaction type

Is the "annotation" a phenotype associated with genotype A?

In the interface we've been testing, the first step is just to select genotype A, not any of the phenotypes annotated to it.

Then, the "phenotype" box doesn't say whether it's meant to collect a phenotype of the first genotype entered, or of the genotype-interacting genotype combination. The fact that it's displayed after the "interaction type" and "interacting genotype" selectors kind of hints that it's for the combination, but it really isn't clear. That needs tidying up and documentation in any case (does that need a new ticket?), but the usefulness of the config file at issue here also depends on the answer - it sounds like a config based on what's chosen as the phenotype for genotype A, not that of the interacting genotype or the combination.

Hmm, does that help at all? Sorry it got so long-winded!

**Comment by vmt25** [comment]
Hi Midori,
Thanks for all of that.

> > CV term (and child terms), annotation, allowed interaction type
> 
> Is the "annotation" a phenotype associated with genotype A?

Now that I read what I wrote, it does not make any sense - i apologize.

Strictly, a config file would need only two columns:
Column A: the CV term/subtree for (on Midori's words) 'a phenotype associated with genotype A'
Column B: Allowed interaction type for CV term/subtree on column A


> the usefulness of the config file at issue here also depends on the answer - it sounds like a config based on what's chosen as the phenotype for genotype A, not that of the interacting genotype or the combination.

You are right.
The interaction type is for a specific phenotype of genotype A - e.g. phenotype 1 of mutant A is suppressed by mutant B, but phenotype 2 of mutant A is not suppressed by mutant B.
Therefore, it would be very useful to select the phenotype before selecting both the interaction type and the interacting genotype. I agree the current configuration can create confusion and needs better labeling and documentation.

**Comment by kimrutherford** [comment]
So would the flow be like this?:

  - select genotype A
  - select genotype B
  - pick a phenotype if this interaction has a phenotype
  - (for PomBase, add conditions)
  - then pick an interaction type from a list of valid types for the phenotype

If that's the case, should we re-order the fields in the dialog to match?

**Comment by vmt25** [comment]
>     * pick a phenotype if this interaction has a phenotype

This is  a bit confusing - do you mean something like ' * pick a phenotype previously annotated to genotype A if genotype A has a phenotype '?


I would prefer step 2 to be the selection of a phenotype of genotype A, precisely to avoid doubts about what is the 'genotype A->phenotype' annotation for this interaction?

I'm unser about the best order of the "* select genotype B' and  '* pick an interaction type from a list of valid types for the phenotype'

**Comment by mah11** [comment]
> I would prefer step 2 to be the selection of a phenotype of genotype A, precisely to avoid doubts about what is the 'genotype A->phenotype' annotation for this interaction?

I'm inclined to agree with this - we want to capture which phenotype of genotype A is affected in the interaction. (But for PomBase we can leave it optional, because even having genotype A is an improvement over only having gene A.)

> I'm unser about the best order of the "* select genotype B' and '* pick an interaction type from a list of valid types for the phenotype'

Me too. I think we might actually want the phenotype of (genotypes A + B combined) here, rather than any phenotype of genotype B alone. This would also be optional - maybe that's what "pick a phenotype if this interaction has a phenotype" refers to?

I could see how it might work either way

- select the (A + B) phenotype, then choose interaction type from a list of valid types for the phenotype

OR

- select an interaction type, then select the the (A + B) phenotype, with ontology constraints that make sense for the interaction type.

Here are two options shown step-by-step:

1. select genotype A
2. find/select phenotype of genotype A
3. select genotype B
4. select the (A + B) phenotype
5. select interaction type (list restricted to types relevant to (A + B) genotype)

OR

1. select genotype A
2. find/select phenotype of genotype A
3. select interaction type
4. select genotype B
5. select the (A + B) phenotype (list restricted to phenotypes relevant to the selected interaction type)

The second feels a bit more intuitive to me, but I suspect I would get used to either.

**Comment by mah11** [comment]
for PomBase also see https://github.com/pombase/curation/issues/2433

**Comment by vmt25** [comment]
> I think we might actually want the phenotype of (genotypes A + B combined) here, rather than any phenotype of genotype B alone.

I am a very confused here...
 
Flybase does not capture the final phenotype of genotypes A and B except for 'synthetic' phenotypes not present in either genotype A or genptype B... which would be done by multiple gene phenotype annotations.

And we don't want to add the phenotype of B here. The reciprocal effect of genotype A on the phenotype of genotype B would be captured by another genetic interaction annotation. 

However, this made me think that it could be useful to have a button to make the reciprocal annotation once an interactions has been created (frequent example: genotypes A and B each have the same phenotype X which is more severe(enhanced) in the double mutant) - it would be good to discuss this on a meeting.


Here is what FlyBase genetic interactions need:

1.     select genotype A
2.     find/select phenotype of genotype A (and associated extensions)
3.     select interaction type (list restricted to types relevant to the phenotype of genotype A) + extensions/conditions from an ontology
4.     select genotype B

This is also my favoured workflow - although other orders could work, provided the labels are very clear.

**Comment by kimrutherford** [comment]
> list restricted to types relevant to the phenotype of genotype A

Hi.  I've done some work on this and I should be able to getting something ready for testing tomorrow.

Do you have a couple of examples of terms that should have a restricted interaction type list, and the interaction types that should be in the list?

**Comment by vmt25** [comment]
> Do you have a couple of examples of terms that should have a restricted interaction type list, and the interaction types that should be in the list?

All terms with limited interactions are on the FBcv/FlyBase_miscellaneous_CV ontology, subtree FBcv:0000347/phenotypic_class)
viable / FBcv:0000349 - no interaction shoiuld be allowed
partially lethal / FBcv:0002015 and child terms - 'suppressible' is allowed; 'suppressible | partially' and 'enhanceable' are not allowed

fertile / FBcv:0000374, male fertile / FBcv:0000376 and female fertile / FBcv:0000377 - no interaction shoiuld be allowed
semi-fertile / FBcv:0000375 and child terms - 'suppressible' is allowed; 'suppressible | partially' and 'enhanceable' are not allowed

We/FlyBase still have to come up with ironclad rules for a few other terms concerning viabiltiy/mortality and fertily/sterility.

**Comment by kimrutherford** [comment]
> All terms with limited interactions are on the FBcv/FlyBase_miscellaneous_CV

Currently the flybase-test Canto is configured to use the FlyBase_anatomy_CV ontology for interactions.  Should we be using FlyBase_miscellaneous_CV instead?  If not, where is FlyBase_miscellaneous_CV used?  I'm a bit confused by the FlyBase CVs.

**Comment by kimrutherford** [comment]
I think the code is finished for this and it's works with my small test configuration.  We should be able to start testing this tomorrow.

**Comment by vmt25** [comment]
> Currently the flybase-test Canto is configured to use the FlyBase_anatomy_CV ontology for interactions. Should we be using FlyBase_miscellaneous_CV instead?

Yes, please!


> where is FlyBase_miscellaneous_CV used?

There will be two different phenotype annotation types in two fields, one for the description of defects which uses FlyBase_miscellaneous_CV, and another describing the anatomy where the phenotpe manifests itself which uses the Flybase anatomy ontology - I managed to implement it on my local canto and is described in #1836 - see a sample output below. Let us leave the duplication for when the interactions are finished and tested.

Given the confusion, we should chat about this next week.

<img width="1006" alt="56203579-742eb380-603d-11e9-9868-08b3188c9295" src="https://user-images.githubusercontent.com/33721108/63503330-2af8df80-c4c7-11e9-9a4f-36a8f546e32b.png">

**Comment by kimrutherford** [comment]
> There will be two different phenotype annotation types in two fields

Thanks for the explanation.  That makes sense.

> Given the confusion, we should chat about this next week.

Yep, I think so too.  I think we're making progress though!

**Comment by vmt25** [comment]
> I think we're making progress though!

Absolutely.

**Comment by kimrutherford** [comment]
I did some more testing of the code that sets the interaction types based on the term and found some bugs.  I think I've fixed things.  The logic has changed so that the first (and only the first) matching configuration is used.  I think that will make configuration more obvious.

To implement this the configuration format has changed a bit.  It now looks something like:

    namespace_term_evidence_codes:
      fly_anatomy.ontology:
        - constraint: "is_a(FBcv:0000349)-is_a(FBcv:0002015)"
          evidence_codes: []

        - constraint: "is_a(FBcv:0002015)"
          evidence_codes:
           - Suppressible

        - constraint: "is_a(FBcv:0000374)-is_a(FBcv:0000375)"
          evidence_codes: []

        - constraint: "is_a(FBcv:0000375)"
          evidence_codes:
           - Suppressible

        - constraint: "is_a(FBcv:0000376)"
          evidence_codes: []

        - constraint: "is_a(FBcv:0000377)"
          evidence_codes: []


I've updated the flybase-test Canto with this change.  Please let me know if you spot any problems.

**Comment by vmt25** [comment]
It is not working.
As you can see, the following annotation should only show 'suppressible' on the list of allowed evidence codes


<img width="828" alt="Screen Shot 2019-09-03 at 13 24 02" src="https://user-images.githubusercontent.com/33721108/64173046-4ca28100-ce4e-11e9-993b-e4fa1a8da9ce.png">



A few things:

1 - Maybe it is not working because
a) the namespace for which the restrictions apply should be 'phenotypic_class' (i.e. they are all FlyBase phenotype annotations instead of FlyBase anatomy annotations) and 
b) the FlyBase anatomy interactions are not working (prob because of the fused ontology with merged terms)
c) So, could you change your code above to?
namespace_term_evidence_codes:
  phenotypic_class:

2 - Would it be possible to have a config file instead of this code?
a) We will have more constraints and envisage some changes to the underlying ontology, so a config file would probably be easier. (We discussed this last week but I don't remember the bottom line - apologies)
b) Other than the namespace, the code reads fine

3 - I assume that not allowing evidence codes will also prevent finalizing an interaction annotation, but could you please confirm?

4 - The interaction evidence codes for both FlyBase phenotype and anatomy interactions should be the ones used by Flybase (all lower case):
evidence_codes:
- suppressible
- suppressible | partially
- enhanceable
- non-suppressible
- non-enhanceable
- non-modified

**Comment by kimrutherford** [comment]
> the namespace for which the restrictions apply should be 'phenotypic_class'
> So, could you change your code above to?
> namespace_term_evidence_codes:
> phenotypic_class:

Thanks.  I made that change and it seems to work.  "partially lethal" does the right thing for me now.

> - Would it be possible to have a config file instead of this code?

That configuration goes in the canto_deploy.yaml file so it can be changed without editing any code.
Here is the current configuration for the flybase-test Canto:
https://curation.pombase.org/kmr44/flybase_canto_deploy-2019-09-05.yaml

> I assume that not allowing evidence codes will also prevent finalizing an interaction annotation,
> but could you please confirm?

Yep, currently you can't complete an interaction annotation without an evidence code (interaction type) so for example it's not possible to make an interaction with the phenotype "viable".

> The interaction evidence codes for both FlyBase phenotype and anatomy interactions should be
> the ones used by Flybase (all lower case):

I've fixed those.

**Comment by ValWood** [comment]
We think the discussing on here was done last week.

The final issues appear to be resolved, but outstanding issues can go in new tickets as this is quite long.
